meadow-client/src/Semantics.purs

```diff
@@ -0,0 +1,1067 @@
+                        , amount :: BigInt
+                        , timeout :: Timeout }
+
+data CommitInfo = CommitInfo { redeemedPerPerson :: M.Map Person BigInt
+                             , currentCommitsById :: M.Map IdCommit CommitInfoRecord 
+                             , expiredCommitIds :: S.Set IdCommit
+                             , timeoutData :: TimeoutData }
+--derive instance eqCommitInfo :: Eq CommitInfo  
+--derive instance ordCommitInfo :: Ord CommitInfo  
+
+type OracleDataPoint = { blockNumber :: BlockNumber
+                       , value :: BigInt }
+
+data State = State { commits :: CommitInfo
+                   , choices :: M.Map WIdChoice Choice
+                   , oracles :: M.Map IdOracle OracleDataPoint
+                   , usedIds :: S.Set IdAction}
+--derive instance eqState :: Eq State  
+--derive instance ordState :: Ord State  
+
+-- Adds a commit identifier to the timeout data map 
+addToCommByTim :: Timeout -> IdCommit -> TimeoutData -> TimeoutData 
+addToCommByTim timeout idCommit timData =
+  case M.lookup timeout timData of
+    Just commSet -> M.insert timeout (S.insert idCommit commSet) timData
+    Nothing -> M.insert timeout (S.singleton idCommit) timData
+
+-- Remove a commit identifier from the timeout data map 
+removeFromCommByTim :: Timeout -> IdCommit -> TimeoutData -> TimeoutData
+removeFromCommByTim timeout idCommit timData =
+  case M.lookup timeout timData of
+    Just commSet -> let newCommSet = S.delete idCommit commSet in
+                    if S.isEmpty newCommSet
+                    then M.delete timeout timData
+                    else M.insert timeout newCommSet timData
+    Nothing -> timData 
+
+-- Add a commit to CommitInfo
+addCommit :: IdCommit -> Person -> BigInt -> Timeout -> State -> State
+addCommit idCommit person value timeout (State state) =
+  State (state { commits = CommitInfo (ci { currentCommitsById =
+                                      M.insert idCommit
+                                               ({ person: person
+                                                , amount: value
+                                                , timeout: timeout })
+                                               ci.currentCommitsById
+                      , timeoutData = addToCommByTim timeout idCommit ci.timeoutData }) })
+  where (CommitInfo ci) = state.commits
+
+-- Return the person corresponding to a commit
+personForCurrentCommit :: IdCommit -> State -> Maybe Person
+personForCurrentCommit idCommit (State state) =
+  case M.lookup idCommit ci.currentCommitsById of
+    Just v -> Just v.person
+    Nothing -> Nothing
+  where (CommitInfo ci) = state.commits
+
+-- Check whether the commit is current (committed not expired)
+isCurrentCommit :: IdCommit -> State -> Boolean
+isCurrentCommit idCommit (State state) =
+  case M.lookup idCommit ci.currentCommitsById of
+    Just _ -> true 
+    Nothing -> false
+  where CommitInfo ci = state.commits
```

Comment:
Actually I'd consider changing this whole thing into:

```
-- Check whether the commit is current (committed not expired)
isCurrentCommit :: IdCommit -> State -> Boolean
isCurrentCommit idCommit (State {commits: (CommitInfo ci)}) =
  M.member idCommit ci.currentCommitsById
```

... or even:

```
-- Check whether the commit is current (committed not expired)
isCurrentCommit :: IdCommit -> State -> Boolean
isCurrentCommit idCommit (State {commits: (CommitInfo {ci.currentCommitsById})}) =
  M.member idCommit currentCommitsById
```

Reply:
Indeed, much better. I had some trouble with the pattern syntax of records, so I gave it up.